src/graphql/graphql-api.php

```diff
@@ -0,0 +1,245 @@
+			// This block doesnt have any inner blocks, so just add it to the flattened blocks. Ensure the parentId is set.
+			if ( ! isset( $inner_block['innerBlocks'] ) ) {
+				$inner_block['parentId'] = $inner_block['parentId'] ?? $filter_options['parentId'];
+				array_push( $flattened_blocks, $inner_block );
+				// This block is a root block, so go through the inner blocks recursively.
+			} elseif ( ! isset( $inner_block['parentId'] ) ) {
+				$inner_blocks_copy = $inner_block['innerBlocks'];
+				unset( $inner_block['innerBlocks'] );
+				$inner_block['parentId'] = $inner_block['parentId'] ?? $filter_options['parentId'];
+				array_push( $flattened_blocks, $inner_block );
+				$flattened_blocks = self::flatten_inner_blocks( $inner_blocks_copy, $filter_options, $flattened_blocks );
+			}
+		}
+
+		return $flattened_blocks;
+	}
+
+	/**
+	 * Register types and fields graphql integration.
+	 *
+	 * @return void
+	 */
+	public static function register_types() {
+		// Register the type corresponding to the attributes of each individual block.
+		register_graphql_object_type(
+			'BlockDataAttribute',
+			[
+				'description' => 'Block data attribute',
+				'fields'      => [
+					'name'  => [
+						'type'        => 'String',
+						'description' => 'Block data attribute name',
+					],
+					'value' => [
+						'type'        => 'String',
+						'description' => 'Block data attribute value',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual inner block, with the above attribute.
+		register_graphql_type(
+			'InnerBlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'         => [
+						'type'        => 'String',
+						'description' => 'ID of the block',
+					],
+					'parentId'   => [
+						'type'        => 'String',
+						'description' => 'ID of the parent for this inner block, if it is an inner block. This will match the ID of the block',
+					],
+					'name'       => [
+						'type'        => 'String',
+						'description' => 'Block name',
+					],
+					'attributes' => [
+						'type'        => [
+							'list_of' => 'BlockDataAttribute',
+						],
+						'description' => 'Block data attributes',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual block, with the above attribute.
+		register_graphql_type(
+			'BlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'          => [
+						'type'        => 'String',
+						'description' => 'ID of the block',
+					],
+					'name'        => [
+						'type'        => 'String',
+						'description' => 'Block name',
+					],
+					'attributes'  => [
+						'type'        => [
+							'list_of' => 'BlockDataAttribute',
+						],
+						'description' => 'Block data attributes',
+					],
+					'innerBlocks' => [
+						'type'        => [ 'list_of' => 'InnerBlockData' ],
+						'description' => 'Flattened list of inner blocks of this block',
```

Comment:
I get why these are flattened, but is the "reparenting" going to be annoying / error-prone? Is the nesting issue severe enough to warrant this approach?

Reply:
The core problem we wanted to solve was that, you should be able to get back the complicated block hierarchy in a post without having any knowledge of the depth involved. This gets us past that, and to ensure that the re-parenting isn't annoying we added a little utility function in the README that could be used. Admittedly that function could be simplified further, but I kept it simple enough to follow. By ensuring that each block gets an id and if applicable, a parent id the likelihood of an error should be minimal (one would hope,  but obviously things can go wrong).

The other approach would be to flatten the innerBlocks alongside all the blocks but that proved to be even more complicated (in terms of code and the outcome) and wouldn't necessarily match the structure of a parent-child in a post with Gutenberg blocks.

So with that in mind, this was the ultimate solution that was picked to allow graphQL to be added on top of the block data api without the innerBlocks being a problem.

Reply:
Ok. I totally get the why, just wondering how big a deal this actually is in practice:

> you should be able to get back the complicated block hierarchy in a post without having any knowledge of the depth involved

If you use fragments in your query, adding [another layer isn't a huge deal](https://github.com/Automattic/vip-go-nextjs-skeleton/blob/b6ad3127e51c5d032255488665956ef1cdd58b9d/graphql/fragments/ContentNode.graphql#L9-L12). And since you are reconstructing the deep tree on the client side, your block parsing code isn't less complex.

Yes, if you add another level beyond what you are querying for, it simply won't be in the response. But adding a `hasInnerBlocks` or `innerBlocksCount` property could allow the queryer to determine if they are "missing out" on blocks.

Another drawback of this approach is that it makes human inspection of the response data much less friendly, possibly impossible. Human debugging is useful and common, especially with the built-in GraphiQL client.

Again, totally understand where this is coming from, just want to make sure the devex has been fully considered.

Reply:
Just an idea: If you use the same type for both inner and outer blocks, you could add a [GraphQL field directive](https://graphql.org/learn/queries/) that allows the queryer to choose whether to flatten or not. If would look like this:

```gql
query FlattenBlocks {
   post(id: "abc123=") {
      blockData(flatten: true) {
         # ...
      }
   }
}

query NestedBlocks {
   post(id: "abc123=") {
      blockData(flatten: false) { # or omit the directive
         # ...
      }
   }
}
```

Reply:
Thanks to the re-write I did of the way the parser's result is transformed into the right graphQL schema, I was able to add in this flatten option. So by default it's set to true because we want the inner blocks to be flattened. Omitting it sets it to true by default. If you set it to false, it will not flatten the inner blocks, and send back the original hierarchy.

So best of both options are available, but we give back the flattened hierarchy by default as thats the best option we want to be used.

Reply:
It is not mentioned in the README

Reply:
Good point, added it in under usage!